extensions/wikidata/tests/src/org/openrefine/wikidata/qa/ConstraintTest.java

```diff
@@ -0,0 +1,65 @@
+package org.openrefine.wikidata.qa;
+
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.wikidata.wdtk.datamodel.helpers.Datamodel;
+import org.wikidata.wdtk.datamodel.interfaces.Claim;
+import org.wikidata.wdtk.datamodel.interfaces.EntityIdValue;
+import org.wikidata.wdtk.datamodel.interfaces.ItemIdValue;
+import org.wikidata.wdtk.datamodel.interfaces.PropertyIdValue;
+import org.wikidata.wdtk.datamodel.interfaces.Reference;
+import org.wikidata.wdtk.datamodel.interfaces.Snak;
+import org.wikidata.wdtk.datamodel.interfaces.SnakGroup;
+import org.wikidata.wdtk.datamodel.interfaces.Statement;
+import org.wikidata.wdtk.datamodel.interfaces.StatementRank;
+import org.wikidata.wdtk.datamodel.interfaces.Value;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ConstraintTest {
+
+    public static ItemIdValue itemIdValue = Datamodel.makeWikidataItemIdValue("Q19474404");
+    public static PropertyIdValue constraintException = Datamodel.makeWikidataPropertyIdValue("P2303");
+    public static Value exceptionValue = Datamodel.makeWikidataItemIdValue("Q7409772");
+    public static PropertyIdValue constraintStatus = Datamodel.makeWikidataPropertyIdValue("P2316");
+    public static Value statusValue = Datamodel.makeWikidataItemIdValue("Q62026391");
+
+    public static Constraint constraint;
+    public static Statement statement;
+
+    @BeforeMethod
+    public void setUp() {
+        Snak snak1 = Datamodel.makeValueSnak(constraintException, exceptionValue);
+        List<Snak> snakList1 = Collections.singletonList(snak1);
+        Snak snak2 = Datamodel.makeValueSnak(constraintStatus, statusValue);
```

Comment:
In general, it would be good to use better variable names: as I reader, I already know this is a `Snak` - this is already written as type. It would be more useful to know how it is used. Is it a main snak? A qualifier? A reference snak?
The same applies to all the variable names you define here.

Reply:
From Wikidata glossary https://www.wikidata.org/wiki/Wikidata:Glossary :
> Snak is a technical term of [Wikibase](https://www.wikidata.org/wiki/Wikidata:Glossary#Wikibase) software which data users are most likely to encounter when [accessing Wikidata through the MediaWiki API](https://www.wikidata.org/wiki/Wikidata:Data_access#MediaWiki_API). It refers to the combination of a [property](https://www.wikidata.org/wiki/Wikidata:Glossary#Property) and either a [value](https://www.wikidata.org/wiki/Wikidata:Glossary#Values) or one of the special cases "no value" and "unknown value". Snaks can be found in [claims](https://www.wikidata.org/wiki/Wikidata:Glossary#Claim) (then they are called main snaks) or in [qualifiers](https://www.wikidata.org/wiki/Wikidata:Glossary#Qualifier) as part of [statements](https://www.wikidata.org/wiki/Wikidata:Glossary#Statement) (then they are called qualifier snaks). E.g., in the statement "Emma Watson was a cast member of [Harry Potter and the Philosopher's Stone](https://www.wikidata.org/wiki/Q102438) in the role of Hermione Granger" there is a main snak "was a cast member of Harry Potter and the Philosopher's Stone" and a qualifier snak "in the role of Hermione Granger".

Reply:
The SnakValidator basically helps validate Snak types?  https://doc.wikimedia.org/Wikibase/master/php/classWikibase_1_1Repo_1_1Validators_1_1SnakValidator.html